python/rmm/rmm.py

```diff
@@ -267,6 +267,31 @@ def get_ipc_handle(ary, stream=0):
     )
 
 
+def make_cupy_use_rmm():
+    """
+    Make CuPy use RMM.
+    Requires the cupy module.
+    """
+    import cupy
+    from rmm._lib.device_buffer import DeviceBuffer
+
+    class RMMMemory(cupy.cuda.memory.BaseMemory):
+        def __init__(self, size):
+            self.size = size
+            self.device_id = cupy.cuda.device.get_device_id()
+            if size > 0:
+                self.rmm_array = DeviceBuffer(size=size)
+                self.ptr = self.rmm_array.ptr
+            else:
+                self.rmm_array = None
+                self.ptr = 0
+
+    def rmm_mem_allocator(bsize):
+        return cupy.cuda.memory.MemoryPointer(RMMMemory(bsize), 0)
```

Comment:
Do you need a new class? Could it just be a DeviceBuffer? (make DeviceBuffer satisfy cupy's expected interface?

Reply:
I don't know if we want to force `DeviceBuffer` to be a subclass of `cupy.cuda.memory.BaseMemory`. Given this is something we'll have to maintain moving forward I'd rather be safer and have a separate class that references a `DeviceBuffer` underneath.

We could imagine needing something similar for Numba, PyTorch, etc. in the future where it would be different subclasses.

Reply:
I think we would mainly need the device ID where the memory was allocated. Is there a good way to get this from RMM?

Reply:
@kkraus14 I thought Python was duck typed, so you wouldn't need to subclass, just provide the methods and members that cupy expects. But it was just a thought and I'm by no means a Python expert.

@jakirkham You tell RMM where to allocate memory. You don't ask it.  You need to initialize RMM with a list of devices on which it should be used. Then when you want to allocate on a given device, you need to `cudaSetDevice()` (CUDA command, not RMM) before calling rmmAlloc().

Reply:
Does `device_buffer` keep track of that information somewhere?

Reply:
I guess we could call [`cudaPointerGetAttributes`]( https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__UNIFIED.html#group__CUDART__UNIFIED_1gd89830e17d399c064a2f3c3fa8bb4390 ) and extract that information.

Reply:
> @kkraus14 I thought Python was duck typed, so you wouldn't need to subclass, just provide the methods and members that cupy expects. But it was just a thought and I'm by no means a Python expert.

CuPy has an explicit `isinstance(obj, cupy.cuda.memory.BaseMemory)` check :/

Reply:
> @kkraus14 I thought Python was duck typed, so you wouldn't need to subclass, just provide the methods and members that cupy expects. But it was just a thought and I'm by no means a Python expert.

CuPy is mostly Cython so a lot of that can go out the window depending on if something is passed into `cdef` functions and whatnot as well.

Reply:
@jakirkham your question 

> Does `device_buffer` keep track of that information somewhere?

It doesn't. At first this question made me worry, because I thought about strange RAII cases where someone loops over devices calling `cudaSetDevice(i)` and creating `device_buffer` on each one, and then the function exits and they go out of scope and get deleted, but all with  the same active CUDA device rather than different ones.

This could be an issue. However here are some reasons we won't change this behavior...

1. `thrust::device_vector` doesn't internally track the device, so `rmm::device_buffer` shouldn't need to.
2. Anyone doing what I described above should not rely on stack variables and RAII, they should probably create `unique_ptr`s to the buffers and explicitly switch back to the appropriate device before destroying them.
3. `cudaFree` doesn't need to be called on the right device -- CUDA can tell from the pointer what device it belongs to. So when RMM is in default mode this won't be an issue. Only `cnmem` will have a problem.